apps/emqx_utils/src/emqx_utils_stream.erl

```diff
@@ -102,6 +104,64 @@ map(F, S) ->
         end
     end.
 
+%% @doc Make a stream by filtering the underlying stream with a predicate function.
+filter(F, S) ->
+    FilterNext = fun FilterNext(St) ->
+        case next(St) of
+            [X | Rest] ->
+                case F(X) of
+                    true ->
+                        [X | filter(F, Rest)];
+                    false ->
+                        FilterNext(Rest)
+                end;
+            [] ->
+                []
+        end
+    end,
+    fun() -> FilterNext(S) end.
+
+%% @doc Consumes the stream and applies the given function to each element.
+foreach(F, S) ->
+    case next(S) of
+        [X | Rest] ->
+            F(X),
+            foreach(F, Rest);
+        [] ->
+            ok
+    end.
+
+%% @doc Drops N first elements from the stream
+-spec drop(non_neg_integer(), stream(T)) -> stream(T).
+drop(N, S) ->
+    DropNext = fun DropNext(M, St) ->
+        case next(St) of
+            [_X | Rest] when M > 0 ->
+                DropNext(M - 1, Rest);
+            Next ->
+                Next
+        end
+    end,
+    fun() -> DropNext(N, S) end.
```

Comment:
Q: Wouldn't it make more sense to turn this function into an eager evaluator, similar to `consume/2`? It shouldn't even change the spec.

Reply:
Depending on whether we put it into the Evaluating or the Combinator section, both variants seem to make sense.

When constructing a stream, it may be unexpected or undesirable to have something evaluated (e.g., when we construct a csv-reading-parsing stream and want to skip the header, we probably do not expect eager reading of a file)

On the other hand, in the actual use case, we just scroll the stream to the desired page, i.e. consume it.

IMO lazy variant is more generic because it wraps eager dropping in just a single call, so the overhead of 
consume(drop(...)) over eager drop is tiny 🤔